src/suffix.c

```diff
@@ -124,55 +128,211 @@ void deleteSuffixTrie(Trie *trie, const char *str, uint32_t len) {
 }
 
-static int processSuffixData(suffixData *data, TrieSuffixCallback callback, void *ctx) {
+static int processSuffixData(suffixData *data, SuffixCtx *sufCtx) {
+  //TrieSuffixCallback callback, void *ctx) {
   if (!data) {
     return REDISMODULE_OK;
   }
   arrayof(char *) array = data->array;
   for (int i = 0; i < array_len(array); ++i) {
-    if (callback(array[i], strlen(array[i]), ctx) != REDISMODULE_OK) {
+    if (sufCtx->callback(array[i], strlen(array[i]), sufCtx->cbCtx, NULL) != REDISMODULE_OK) {
       return REDISEARCH_ERR;
     }
   }
   return REDISMODULE_OK;
 }
 
-static int recursiveAdd(TrieNode *node, TrieSuffixCallback callback, void *ctx) {
+static int recursiveAdd(TrieNode *node, SuffixCtx *sufCtx) {
   if (node->payload) {
     size_t rlen;
-    // printf("nodestr %s len %d rlen %ld", runesToStr(node->str, node->len, &rlen), node->len, rlen);
     suffixData *data = Suffix_GetData(node);
-    processSuffixData(data, callback, ctx);
+    if (processSuffixData(data, sufCtx) != REDISMODULE_OK) {
+      return REDISMODULE_ERR;
+    }
   }
   if (node->numChildren) {
     TrieNode **children = __trieNode_children(node);
     for (int i = 0; i < node->numChildren; ++i) {
-      // printf("child %d ", i);
-      if (recursiveAdd(children[i], callback, ctx) != REDISMODULE_OK) {
+      if (recursiveAdd(children[i], sufCtx) != REDISMODULE_OK) {
         return REDISMODULE_ERR;
       }
     }
   }
   return REDISMODULE_OK;
 }
 
-void Suffix_IterateContains(TrieNode *n, const rune *str, size_t nstr, bool prefix,
-                            TrieSuffixCallback callback, void *ctx) {
-  if (prefix) {
+void Suffix_IterateContains(SuffixCtx *sufCtx) {
+  if (sufCtx->type == SUFFIX_TYPE_CONTAINS) {
     // get string from node and children
-    TrieNode *node = TrieNode_Get(n, str, nstr, 0, NULL);
+    TrieNode *node = TrieNode_Get(sufCtx->root, sufCtx->rune, sufCtx->runelen, 0, NULL);
     if (!node) {
       return;
     }
-    recursiveAdd(node, callback, ctx);
-  } else {
+    recursiveAdd(node, sufCtx);
+  } else if (sufCtx->type == SUFFIX_TYPE_SUFFIX) {
     // exact match. Get strings from a single node
-    TrieNode *node = TrieNode_Get(n, str, nstr, 1, NULL);
+    TrieNode *node = TrieNode_Get(sufCtx->root, sufCtx->rune, sufCtx->runelen, 1, NULL);
     suffixData *data = Suffix_GetData(node);
     if (data) {
-      processSuffixData(data, callback, ctx);
+      processSuffixData(data, sufCtx);
+    }
+  }
+}
+
+
+/***********************************************************************************
+*                                    Wildcard                                      *
+************************************************************************************/
+
+/* Breaks wildcard at '*'s and find the best token to get iterate the suffix trie  */
+int Suffix_ChooseToken(const char *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
+  int runner = 0;
+  int i = 0;
+  int init = 0;
+  while (i < len) {
```

Comment:
Can we use something like `strstr` ?

Reply:
We need a version for `rune` as well, and I would rather have similar versions.